Add general guidelines for writing test scripts regarding RFC 2119 Ke… (#154)
SovereignCloudStack/docs: contributor-docs/operations/tests/rfc2119-keyword-test-guide.md

---
title: SCS RFC2119 Keyword Test Guide
type:
status: Draft
track: Global
---

## Introduction

The process of test creation and standards definition in Sovereign Cloud Stack (SCS)
involves developing and validating tests that align with the specific requirements and
standards set by SCS. This entails a comprehensive understanding of the SCS framework,
which is built on open-source components and aims to provide interoperable and sovereign
cloud services. In creating tests for SCS, it's essential to consider the various aspects
of cloud technology, including infrastructure, services, and the integration of these
components. The tests should be designed to ensure that the SCS standards are met, which

> includes adhering to specified requirement levels as indicated by [RFC2119 keywords](https://datatracker.ietf.org/doc/html/rfc2119).

Reviewer: The introduction is good but also very generic. In my opinion, it does not yet answer important questions: Why is it needed/why should I read it? Who should read it (or to whom is it important)?

SCS-compliant standard documents.

### Web Server Installation Standard

#### Purpose

This document provides a concise set of requirements for installing a secure web server.

#### Requirements

* Web Server Software: Apache HTTP Server 2.4.x or Nginx 1.18.x **MUST** be installed. No
other web server software versions or types are permitted.

* Operating System: The server **MUST** run on an OS that receives current security updates.
Unsupported versions **MUST NOT** be used.

* TLS Configuration: TLS 1.2 or higher **MUST** be enabled for all connections. SSL and
earlier TLS versions **MUST NOT** be used.

* Security Patches: The web server and OS **MUST** be kept up-to-date with security patches.
Critical patches **SHOULD** be applied within 48 hours of release.

* Firewall Setup: The firewall **SHOULD** be configured to allow only necessary ports,
such as 80 (HTTP) and 443 (HTTPS). All unnecessary ports **SHOULD NOT** be open.

* Backup Plan: Regular backups of the web server data **MAY** be performed, and it is
recommended that backup integrity checks are conducted to ensure recoverability.

#### Compliance

> Failure to comply with these requirements results in the web server being

Reviewer: ```suggestion
Failure to comply with these requirements results in the web server installation being
```

As I understand it, the example standard is about the installation (and configuration) of the webserver and its environment and not the webserver itself.

Author: Here I disagree, if the installation is not followed strictly and therefore not compliant, the webserver itself will be not compliant with these rules.

of cloud technology, including infrastructure, services, and the integration of these
components. The tests should be designed to ensure that the SCS standards are met, which
includes adhering to specified requirement levels as indicated by [RFC2119 keywords](https://datatracker.ietf.org/doc/html/rfc2119).

## 1. Understanding SCS Standards and RFC2119 Keywords

SCS standards use RFC2119 keywords like:

- **MUST** (same as **REQUIRED**, **SHALL**),
- **MUST NOT** (same as **SHALL NOT**),
- **SHOULD** (same as **RECOMMENDED**),
- **SHOULD NOT** (same as **NOT RECOMMENDED**),
- **MAY** (same as **OPTIONAL**)

to define requirements for SCS-compatible IaaS and KaaS resources. These standards are crucial for ensuring interoperability and sovereignty in cloud services, as they are built on open-source components such as Kubernetes.

To ensure readability and comprehensibility, only the main keywords **MUST**, **MUST NOT**, **SHOULD**, **SHOULD NOT** and **MAY** are referred to below.

## 2. Interpreting RFC2119 Keywords in the Context of SCS

- **MUST and MUST NOT**: These are binary and straightforward to test. Compliance or

>   non-compliance directly affects the standard meeting. For visualization, fulfilled

Reviewer: ```suggestion
  non-compliance directly affects whether the standard is satisfied. For visualization, fulfilled
```

(To me, *standard meeting* sounds like something were people meet.)

Author: changed

  requirements could be marked in green, unfulfilled ones in red.
- **SHOULD and SHOULD NOT**: These represent recommendations. While fulfilling these do
  not directly affect standard compliance, it is advised for future-proofing against
  potential standard evolutions. Non-compliance could be visualized in yellow, possibly
  with a valid reason for non-fulfillment.
- **MAY**: This is more flexible and could be used to provide additional
  information or best practices. Their impact on compliance is not as direct as **MUST** or
  **SHOULD** categories. These can be marked as blue.

## 3. Channels for Output in Test Scripts

- Align the RFC2119 keywords with specific channels in test scripts:
  - **MUST** and **MUST NOT** lead to the ERROR channel for failed tests
  - **SHOULD** and **SHOULD NOT** lead to the WARNING channel, the test is still passed

>   - **MAY** to leads to the INFO channel, the test is still passed

Reviewer: The output channels seem to come out of "nowhere", have you considered introducing them briefly beforehand? And also why they are used (this might be good to know).

Something along the lines of 

> In test scripts, we use *channels* to convey information of different importance to the user. These channels are based on the ubiquitous and de-facto standard logging levels of common logging libraries in scripting and programming languages, in particular Python: DEBUG, INFO, WARNING, ERROR, and CRITICAL.

Furthermore, because the first enumeration level just has a single item ("- Align the RFC2119 keywords ..."), I think it should be "flattened".

Author: Added the explanation. 
Changed the enumeration level.

        logger.debug(f"var1 evaluated: {var1 > 0 and var1 < 100}")

        # example of a warning case (equivalent to SHOULD and SHOULD NOT)
        if var2 < 10:
            logger.warning("var2 should be at least 10 for optimal performance")
        elif var2 > 50:
            logger.warning("var2 should not exceed 50 for optimal performance")

        # add debug log to confirm evaluation of var2
        logger.debug(f"var2 evaluated: {var2 > 10 and var2 < 50}")

        # example of an informative case (equivalent to MAY)
        if var1 + var2 < 100:
            logger.info("Combination is within recommended range")
        else:
            logger.info("Combination exceeds the recommended range but may still proceed")

        # debug log before completing the process
        logger.debug(f"Final evaluation before completion: var1 + var2 = {var1 + var2}")

        # success log
        logger.debug("Process completed successfully")

    except ValueError as e:
        # debug log before raising a critical error
        logger.debug("Encountered a critical error about to log", exc_info=True)
        logger.critical(f"Process failed: {e}", exc_info=True)

> ```

Reviewer: IMHO somewhere in this document it should be documented that the exit code of the conformance test script must be non-zero if we have ERROR or CRITICAL messages.

Author: I added this as well.

## 2. Interpreting RFC2119 Keywords in the Context of SCS

- **MUST and MUST NOT**: These are binary and straightforward to test. Compliance or
  non-compliance directly affects whether the standard is satisfied. For visualization, fulfilled
  requirements could be marked in green, unfulfilled ones in red.
- **SHOULD and SHOULD NOT**: These represent recommendations. While fulfilling these do
  not directly affect standard compliance, it is advised for future-proofing against
  potential standard evolutions. Non-compliance could be visualized in yellow, possibly
  with a valid reason for non-fulfillment.
- **MAY**: This is more flexible and could be used to provide additional
  information or best practices. Their impact on compliance is not as direct as **MUST** or
  **SHOULD** categories. These can be marked as blue.

## 3. Channels for Output in Test Scripts

In test scripts, different channels are used to convey information of different importance to the user. These channels are based on the ubiquitous and de-facto standard logging levels of common logging libraries in scripting and programming languages, in particular Python: DEBUG, INFO, WARNING, ERROR and CRITICAL.

> Alignement of the RFC2119 keywords with specific channels in test scripts:

Author: done

- **MUST** and **MUST NOT** lead to the ERROR channel for failed tests
- **SHOULD** and **SHOULD NOT** lead to the WARNING channel, the test is still passed
- **MAY** leads to the INFO channel, the test is still passed

However, the CRITICAL level is not directly associated with any of the RFC2119 keywords; it is instead used to signify that a test was unable to complete or was interrupted due to various issues, such as runtime failures.

## 4. Compliance and Test Passing Criteria

A test is considered to pass if and only if it doesn't produce any messages on the
ERROR nor the CRITICAL channel. The presence of warnings or informational items from
**SHOULD** and **MAY** categories do not directly impact the pass/fail status but is
important for overall quality and compatibility with future standards.

That means a test can have one of three results:

- fail: if it has at least one message on ERROR
- DNF (did not finish): if it has no message on ERROR, but at least one on CRITICAL
- pass: otherwise

In addition, a test **MUST** exit with a non-zero exit code (e.g., via `sys.exit(…)`) if there are any ERROR or CRITICAL messages, thus signaling a failure to meet a standard.

> It is crucial to determine and explicitly state within this document the expected conventions for logging output, specifically regarding the redirection of all channel outputs to standard error (**stderr**), which aligns with Python's default logging behavior. The format which **MUST** be used is **CHANNEL(MESSAGE)**, where **CHANNEL** represents the log level (e.g., DEBUG, INFO, WARNING, ERROR, CRITICAL) and **MESSAGE** encapsulates the actual log message.

Reviewer: > It is crucial to determine and explicitly state within this document the expected conventions for logging output

This sounds like a prompt for the reader to determine the conventions and should be shortened to something like "the scripts are expected to ...".

Instead of `CHANNEL(MESSAGE)` the actual format we use is `CHANNEL: MESSAGE`. Example:

    INFO: Checking cluster specified by context 'my-cluster' in /path/to/.kube/config.

Author: done

> It is crucial to determine and explicitly state within this document the expected conventions for logging output. In test scripts it is a **MUST** to redirect all channel outputs to standard error (**stderr**), which aligns with Python's default logging behavior. The format which **MUST** be used is **CHANNEL: MESSAGE**, where **CHANNEL** represents the log level (e.g., DEBUG, INFO, WARNING, ERROR, CRITICAL) and **MESSAGE** encapsulates the actual log message.

Reviewer: ```suggestion
In test scripts it is a **MUST** to redirect all channel outputs to standard error (**stderr**), which aligns with Python's default logging behavior. The format which **MUST** be used is **CHANNEL: MESSAGE**, where **CHANNEL** represents the log level (e.g., DEBUG, INFO, WARNING, ERROR, CRITICAL) and **MESSAGE** encapsulates the actual log message.
```

Author: done

> For instance, when logging an ERROR regarding a Kubernetes cluster, the format on the command line should be like this:

Author: done

standards are crucial for ensuring interoperability and sovereignty in cloud services,
as they are built on open-source components such as Kubernetes.

## 2. Interpreting RFC2119 Keywords in the Context of SCS

* **MUST and MUST NOT**: These are binary and straightforward to test. Compliance or
non-compliance directly affects the standard meeting. For visualization, fulfilled
requirements could be marked in green, and unfulfilled ones in red.
* **SHOULD and SHOULD NOT**: These represent recommendations. While fulfilling these do
not directly affect standard compliance, it is advised for future-proofing against
potential standard evolutions. Non-compliance could be visualized in yellow, possibly
with a valid reason for non-fulfillment.
* **MAY**: This is more flexible and could be used to provide additional
information or best practices. Their impact on compliance is not as direct as "MUST" or
"SHOULD" categories. These can be marked as blue.

## 3. Channels for Output in Test Scripts

* Align the RFC2119 keywords with specific channels in test scripts:
  * "MUST" and "MUST NOT" should lead to errors
  * "SHOULD" and "SHOULD NOT" to warnings

>   * "MAY" to notes or hints.

Reviewer 2: We should be more specific about these channels.
I would assume the relevant output to happen on `stderr`, and the channel is indicated by the first word on the line, so the line would start with `ERROR` (ideally all uppercase), for instance, followed by a colon or a whitespace character.
the channel for "SHOULD" would be `WARNING`,
the channel for "MAY" would be `INFO`,
and there would be a fourth channel called `CRITICAL` for circumstances where the test cannot be performed.

Reviewer 2: The final part cannot be stressed enough: if the test cannot be performed for whatever reason, a message MUST be emitted on CRITICAL, and this means: definitely catch all exceptions!

## 4. Compliance and Test Passing Criteria

> * Emphasize that for a test to pass, it must meet all the "MUST" requirements without 

Reviewer 2: The beginning reads like a prompt for someone who should actually write the text -- did you by any change forget to run it through an LLM?

I would write it like so:

> A test is considered to pass if and only if it doesn't produce any messages on the ERROR nor the CRITICAL channel.

Or, maybe even like so:

> A test can have one of three results:
>
>  - fail: if it has at least one message on ERROR
>  - DNF (did not finish): if it has no message on ERROR, but at least one on CRITICAL
>  - pass: otherwise
>
> The messages on the other channels serve informational purposes, for instance, how well the test subject might perform with future versions of the standard.

Author: Agreed, for now I reworded the first sentence.

errors. The presence of warnings or informational items from "SHOULD" and "MAY" 
categories does not directly impact the pass/fail status but is important for overall 
quality and compatibility with future standards.

## 5. Conclusion and Continuous Improvement

* Summarize the importance of adhering to these guidelines for maintaining compliance 
and adaptability within the SCS framework. Highlight the need for ongoing updates and 
revisions to the document to align with evolving standards and best practices in cloud 

> infrastructure and services.

Reviewer 2: This again reads like a prompt.

Author: Agreed, for now I excluded this section since for me it does not really makes sense here anyway.
